### ISSUE TITLE
Bug Report: Training Loss Computed Using Previous Epoch’s Weights

### ISSUE THREAD

**Issue by MahdiTavakol** [MahdiTavakol/ModernCppProjects#6]
### 📘 Description
The **training loss** reported during the `fit()` loop is calculated **before** the network weights are updated, whereas the **validation loss** is calculated **after** all updates are applied.  
This causes the training loss to appear **higher than the validation loss**, even when the model is learning correctly.

During each epoch:
- `trainBatches(..., doBack=true)` computes loss **before** calling `backwardBatch()` and `updateBatch()`.
- `trainBatches(..., doBack=false)` (validation) runs afterward, using the **updated weights**.

| Metric | Weights Used | Timing |
|---------|--------------|---------|
| Training Loss | `W_i` (before updates) | During epoch |
| Validation Loss | `W_{i+1}` (after updates) | After epoch |

Hence, the two losses are not directly comparable.

---

### 🔍 Steps to Reproduce
1. Run `NeuralNetwork::fit()` with any dataset (even without dropout or regularization).  
2. Observe that the training loss is consistently higher than the validation loss.  
3. Inspect the training loop:
   ```cpp
   trainBatches(..., doBack=true);   // computes loss before updates
   trainBatches(..., doBack=false);  // computes val loss after updates
